test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs

```diff
@@ -0,0 +1,288 @@
+            (coinSelectionUnitTest
+                Fixture
+                { maxNumOfInputs = 2
+                , utxoInputs = [1,2,10,6,5]
+                , txOutputs = 11 :| [1]
+                , expectedResult = Left $ MaximumInputsReached 2
+                })
+
+    describe "Coin selection properties : LargestFirst algorithm" $ do
+        it "forall (UTxO, NonEmpty TxOut), \
+           \ running algorithm twice yields exactly the same result"
+            (property propDeterministic)
+        it "forall (UTxO, NonEmpty TxOut), \
+           \ there's at least as many selected inputs as there are requested outputs"
+            (property propAtLeast)
+        it "forall (UTxO, NonEmpty TxOut), for all selected input, \
+           \ there's no bigger input in the UTxO that is not already in the selected inputs."
+            (property propInputDecreasingOrder)
+
+{-------------------------------------------------------------------------------
+                 Properties and unit test generic scenario
+-------------------------------------------------------------------------------}
+
+data Fixture = Fixture
+    { maxNumOfInputs :: Word64
+    , utxoInputs :: [Word64]
+    , txOutputs :: NonEmpty Word64
+    , expectedResult :: Either CoinSelectionError [Word64]
+    } deriving Show
+
+
+coinSelectionUnitTest
+    :: Fixture
+    -> Expectation
+coinSelectionUnitTest (Fixture n utxoCoins txOutsCoins expected) = do
+    (utxo,txOuts) <- setup
+
+    result <- runExceptT $ do
+        CoinSelection inps _ _ <-
+            largestFirst (defaultCoinSelectionOptions n) utxo txOuts
+        return $ map (getCoin . coin . snd) inps
+
+    case expected of
+        Left err ->
+            result `shouldBe` (Left err)
+        Right expectedCoinsSel ->
+            result `shouldBe` (return expectedCoinsSel)
+    where
+        setup :: IO (UTxO, NonEmpty TxOut)
+        setup = do
+            ins <- generate $ vectorOf (L.length utxoCoins) arbitrary
+            addrs <- generate $ vectorOf (L.length utxoCoins) arbitrary
+            let utxo = UTxO $ Map.fromList
+                    $ L.zip ins
+                    $ L.zipWith TxOut addrs
+                    $ map Coin utxoCoins
+            txOutsAddrs <- generate $ vectorOf (L.length txOutsCoins) arbitrary
+            let txOuts = NE.zipWith TxOut (NE.fromList txOutsAddrs)
+                    $ NE.map Coin txOutsCoins
+            pure (utxo, txOuts)
+
+
+propDeterministic
+    :: CoveringCase
+    -> Property
+propDeterministic (CoveringCase (utxo, txOuts)) = do
+    let opts = defaultCoinSelectionOptions 100
+    let resultOne = runIdentity $ runExceptT $ largestFirst opts utxo txOuts
+    let resultTwo = runIdentity $ runExceptT $ largestFirst opts utxo txOuts
+    resultOne === resultTwo
+
+propAtLeast
+    :: CoveringCase
+    -> Property
+propAtLeast (CoveringCase (utxo, txOuts)) =
+    isRight selection ==> let Right s = selection in prop s
```

Comment:
So here I've required that selection is Right as a precondition because our property only matters for valid selections.

Reply:
beautiful :100: